### ISSUE TITLE
About inference of combinators taking records (type, partial ...)

### ISSUE THREAD

**Issue by safareli** [gcanti/io-ts#541]
## 📖 Documentation

I was looking at type of Decode.type:
```ts
export declare const type: <A>(
  properties: { [K in keyof A]: Decoder<unknown, A[K]> }
) => Decoder<unknown, { [K_1 in keyof A]: A[K_1] }>
```

bu the first glance use of `{ [K_1 in keyof A]: A[K_1] }` seams strange, as `A` would have worked as well:
```ts
export declare const type: <A>(
  properties: { [K in keyof A]: Decoder<unknown, A[K]> }
) => Decoder<unknown, A>
```

downside of current approach is illustrated here:
```ts
import * as D from '../src/Decoder'
interface Person {
  name: string
}
const Person = D.type<Person>({
  name: D.string
})
/*
inferred as
const Person: D.Decoder<unknown, {
    name: string;
}>
should be
const Person: D.Decoder<unknown, Person>
*/
```

And workaround for this is to add an explicit annotation:
```ts
const Person: _.Decoder<unknown, Person> = _.type<Person>({
  name: _.string
})
```

I discussed this with @gcanti and as he said "the weird signature coming from  [here](https://github.com/gcanti/io-ts/commit/7c058b828592f7bcb7a885a36a7cccbb5316cf38) where I renamed type to ktype and then re-added type without cleaning its signature".

Unfortunately @gcanti has tried to change the type signature to:
`<A>(properties: { [K in keyof A]: Decoder<unknown, A[K]> }): Decoder<unknown, A>`
but that made the inferred type worse:
```ts
export const d = _.type({
  a: _.type({
    b: _.number
  })
})
/*
inferred as const d: _.Decoder<unknown, {
    a: {
        b: ...;
    };
}>
instead of
const d: _.Decoder<unknown, {
    a: {
        b: number;
    };
}>
*/
```
To be precise, inference works fine, it's just when you hover you don't see the fully expanded type.

to recap:

-  `{ [K_1 in keyof A]: A[K_1] }`
   - Pro: inferred type is expanded
   - Cons: inferred type is still expanded even if you specify the generic type to some interface for which you expect type to not be expanded.
- `A`
   - Pro : Interfaces are not expanded
   - Cons: inferred types are also not expanded.
